File: apps/server/src/models/news/news.ts
import mongoose, { Schema, Document } from "mongoose";
import mongoose_delete from "mongoose-delete";
import slugify from "slugify";

/* Interface for News */
export interface INews extends Document {
  user: mongoose.Schema.Types.ObjectId;
  slug: string;
  title: string;
  category: mongoose.Schema.Types.ObjectId;
  time_read: number;
  mainImage: string;
  content: string;
  tags: mongoose.Schema.Types.ObjectId[];
  views: number;
  status: string;
  message?: string;
  approved: boolean;
  created_at?: Date;
  updated_at?: Date;
}

/* News Schema */
const NewsSchema = new Schema<INews>(
  {
    user: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "user",
      required: true,
    },
    slug: {
      type: String,
      unique: true,
      required: true,
      set: (slug: string) => setSlug(slug),
    },
    title: {
      type: String,
      unique: true,
      required: true,
    },
    category: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "newsCategory",
      required: true,
    },
    time_read: {
      type: Number,
      required: true,
    },
    mainImage: {
      type: String,
      required: true,
    },
    content: {
      type: String,
      required: true,
    },
    tags: {
      type: [mongoose.Schema.Types.ObjectId],
      ref: "newsTag",
      required: true,
    },
    views: {
      type: Number,
      required: true,
      default: 0,
    },
    status: {
      type: String,
      required: true,
      default: "process",
    },
    message: {
      type: String,
      required: false,
    },
    approved: {
      type: Boolean,
      required: true,
      default: false,
    },
  },
  {
    timestamps: {
      createdAt: "created_at",
      updatedAt: "updated_at",
    },
    toJSON: { getters: true, virtuals: true },
    toObject: { virtuals: true },
  }
);

/* Slug function */
function setSlug(slug: string): string {
  return slugify(slug, { lower: true, locale: "id" });
}

/* Apply mongoose-delete plugin */
NewsSchema.plugin(mongoose_delete, { overrideMethods: "all" });

/* Model Definition */
const News = mongoose.model<INews>("news", NewsSchema);

export default News;

File: apps/backend/models/news.go
package models

import (
	"time"

	"go.mongodb.org/mongo-driver/bson/primitive"
)

type News struct {
	ID        primitive.ObjectID   `json:"id" bson:"_id,omitempty"`
	User      primitive.ObjectID   `json:"user" bson:"user"`
	Slug      string               `json:"slug" bson:"slug"`
	Title     string               `json:"title" bson:"title"`
	Category  primitive.ObjectID   `json:"category" bson:"category"`
	TimeRead  int                  `json:"time_read" bson:"time_read"`
	MainImage string               `json:"mainImage" bson:"mainImage"`
	Content   string               `json:"content" bson:"content"`
	Tags      []primitive.ObjectID `json:"tags" bson:"tags"`
	Views     int                  `json:"views" bson:"views"`
	Status    string               `json:"status" bson:"status"`
	Message   string               `json:"message,omitempty" bson:"message,omitempty"`
	Approved  bool                 `json:"approved" bson:"approved"`
	Deleted   *bool                `json:"deleted,omitempty" bson:"deleted,omitempty"`
	CreatedAt time.Time            `json:"created_at,omitempty" bson:"created_at,omitempty"`
	UpdatedAt time.Time            `json:"updated_at,omitempty" bson:"updated_at,omitempty"`
}
